vadl/main/vadl/ast/SymbolTable.java

```
@@ -247,6 +262,15 @@ Macro getMacro(String name) {
     return null;
   }
 
+  /**
+   * This will add an error to {@link #errors} if the identifier couldn't be resolved
+   * to a node of the given type.
+   * In this case it will return null, so the user must check the result before continuing.
+   *
+   * @param usage the identifier that should be resolved
+   * @param type  the type that the resolved node must have
+   * @return the resolved node, or null if it could not be resolved with the given type
+   */
   // FIXME: I don't like how it's called require but still returns null
   <T extends Node> @Nullable T requireAs(IsId usage, Class<T> type) {
```

**question:** Now that you have had more contact with the symbol table, do you think it makes sense to keep this method or would it be better to only rely on `findAs`?

Also should that method be private, because if called after the symbolresolver the errors will get added to the symboltable but no other code checks if there are new errors in the symboltable, right?

---

I think it does make sense, otherwise every caller would have to add an error manually. Making it private is a good idea, however, the Parser needs the method to find ISA definitions (this is a special case, as macros from super ISAs are available in a sub-ISA). Maybe for the parser we should add a new method that makes clear that this is only used by the parser. 

All other usages of this method can be replaced by a call to the `target()` of the identifier.

---

Oh I guess, and if the parser doesn't find it it also should throw an error...
Well then we might keep it but still unhappy that after the symbolresolver the method is still visible but shouldn't be used.

---

I now refactor the symbol table, so almost all methods are private, which should hopefully prevent misuse.